README
ShiftJIS::X0213::MapUTF version 0.31
================

NAME

ShiftJIS::X0213::MapUTF - conversion
between Shift_JIS-2004/Shift_JISX0213 and Unicode

SYNOPSIS

    use ShiftJIS::X0213::MapUTF;

    # for Shift_JIS-2004 (specified by JIS X 0213:2004)
    $utf16be_string  = sjis2004_to_utf16be($sjis2004_string);
    $sjis2004_string = utf16be_to_sjis2004($utf16be_string);

    # for Shift_JISX0213 (specified by JIS X 0213:2000)
    $utf16be_string  = sjis0213_to_utf16be($sjis0213_string);
    $sjis0213_string = utf16be_to_sjis0213($utf16be_string);

DESCRIPTION

This module provides the following functions:

    sjis2004_to_utf8()    converts Shift_JIS-2004 to UTF-8.
    sjis2004_to_utf16le() converts Shift_JIS-2004 to UTF-16LE.
    sjis2004_to_utf16be() converts Shift_JIS-2004 to UTF-16BE.
    sjis2004_to_utf32le() converts Shift_JIS-2004 to UTF-32LE.
    sjis2004_to_utf32be() converts Shift_JIS-2004 to UTF-32BE.
    sjis2004_to_unicode() converts Shift_JIS-2004 to Unicode[*].

    utf8_to_sjis2004()    converts UTF-8      to Shift_JIS-2004.
    utf16le_to_sjis2004() converts UTF-16LE   to Shift_JIS-2004.
    utf16be_to_sjis2004() converts UTF-16BE   to Shift_JIS-2004.
    utf32le_to_sjis2004() converts UTF-32LE   to Shift_JIS-2004.
    utf32be_to_sjis2004() converts UTF-32BE   to Shift_JIS-2004.
    unicode_to_sjis2004() converts Unicode[*] to Shift_JIS-2004.

    utf16_to_sjis2004()   converts UTF-16  to Shift_JIS-2004.
    utf32_to_sjis2004()   converts UTF-32  to Shift_JIS-2004.

    sjis0213_to_utf8()    converts Shift_JISX0213 to UTF-8.
    sjis0213_to_utf16le() converts Shift_JISX0213 to UTF-16LE.
    sjis0213_to_utf16be() converts Shift_JISX0213 to UTF-16BE.
    sjis0213_to_utf32le() converts Shift_JISX0213 to UTF-32LE.
    sjis0213_to_utf32be() converts Shift_JISX0213 to UTF-32BE.
    sjis0213_to_unicode() converts Shift_JISX0213 to Unicode[*].

    utf8_to_sjis0213()    converts UTF-8      to Shift_JISX0213.
    utf16le_to_sjis0213() converts UTF-16LE   to Shift_JISX0213.
    utf16be_to_sjis0213() converts UTF-16BE   to Shift_JISX0213.
    utf32le_to_sjis0213() converts UTF-32LE   to Shift_JISX0213.
    utf32be_to_sjis0213() converts UTF-32BE   to Shift_JISX0213.
    unicode_to_sjis0213() converts Unicode[*] to Shift_JISX0213.

    utf16_to_sjis0213()   converts UTF-16  to Shift_JISX0213.
    utf32_to_sjis0213()   converts UTF-32  to Shift_JISX0213.

    [*] .. Perl's internal format. see perlunicode.pod.

DEPENDENCIES

This module requires Perl 5.6.1 or later.
(sorry, cannot be built with Perl 5.6.0!)

INSTALLATION

You need a C compiler to build this module.
To install this module type the following:

    perl Makefile.PL
    make
    make test
    make install

COPYRIGHT AND LICENSE

SADAHIRO Tomoyuki <SADAHIRO@cpan.org>

Copyright(C) 2002-2007, SADAHIRO Tomoyuki. Japan. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

SEE ALSO

* JIS X 0213:2000/Amd1:2004

    7-bit and 8-bit double byte coded extended KANJI sets
    for information interchange



Changes
Revision history for Perl module ShiftJIS::X0213::MapUTF.

0.31  Mon Jun  4 22:26:05 2007
    - mkheader, unilen() - avoid unpack('C*') on unicode.

0.30  Mon Mar 21 10:14:02 2005
    - META.yml is added.
    - In s10heavy.t and t10heavy.t, UTF-8 strings to be compared are encoded
      without the help of perl's "utf8" (as perl-current breaks these tests).

0.21  Fri Mar 19 22:29:28 2004
    - The encoding name will be corrected from Shift_JIS-2003
      to Shift_JIS-2004 next month (as I have been informed).
      (see also KAISETSU 2.12 in JIS X 0213:2004.)
    * replaced sjis2003 by sjis2004, and Shift_JIS-2003 by Shift_JIS-2004.
      (Function names are also corrected.)

0.20  Sun Feb 29 21:00:05 2004
    - Enhancement for JIS X 0213:2004 (Amendment 1), whose repertoire is
      ten chracters larger than JIS X 0213:2000.
    * New functions named sjis2003_to_u* and u*_to_sjis2003 support
      Shift_JIS-2003 specified by JIS X 0213:2004.
    * Old functions named sjis0213_to_u* and u*_to_sjis0213
      continue to support Shift_JISX0213 specified by JIS X 0213:2000,
      but mapping for SJIS 0xFC5A is fixed after JIS X 0213:2004.

0.13  Sat Oct 18 12:01:25 2003
    - pod: added link to JISC.

0.12  Sat Oct 11 22:35:36 2003
    - manipulation of illegal ShiftJIS strings: SJIS_CALLBACK coderef are
      called with two parameters (like UNICODE_CODEREF).
    - Reserved bytes (0x80, 0xA0, 0xFD..0xFF) are no longer
      a single-byte character.
    - Internal tweak in XSUB: no longer relies on a string to be
      terminated with '\0'.
    - added utf16_to_sjis0213() and utf32_to_sjis0213().
      [but not yet sjis0213_to_utf16() and sjis0213_to_utf32()]
    - modified t04xncr.t avoiding a non-character.

0.11  Sun Nov 03 19:32:21 2002
    - fix a momery leak when the process died in a callback.
      like: eval { utf32be_to_sjis0213(sub { die }, "\x00"); };

0.10  Sat Nov 02 21:12:23 2002
    - manipulation of illegal Unicode strings: u*_to_sjis0213 functions
      may call UNICODE_CALLBACK coderef with two parameters.
    - UTF-8, UTF-32(BE|LE) are supported.
    - tests are improved.

0.02  Sun Jun 30 19:45:16 2002
    - change the mapping for 0x81D4, 0x81D5, 0xFC5A in Shift_JISX0213
      according to Shibano's JIS KANJI JITEN, the revised edition.
    - add Supplements.pm

0.01  Sun Jun 16 18:45:16 2002
    - original version; created by h2xs 1.19


MapUTF/Supplements.pm
package ShiftJIS::X0213::MapUTF::Supplements;

use strict;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK);
use vars qw(%Supplements);

require Exporter;
@ISA = qw(Exporter);

@EXPORT = qw(to_sjis0213_supplements to_sjis2004_supplements);
@EXPORT_OK = qw(%Supplements);

$VERSION = '0.31';

%Supplements = (
  0x00B5,	# MICRO SIGN (ISO/IEC 8859-1 11/05)
    "\x83\xCA",	# GREEK SMALL LETTER MU (U+03BC)

  0x0110,	# LATIN CAPITAL LETTER D WITH STROKE (ISO/IEC 8859-2 13/00)
    "\x85\x66",	# LATIN CAPITAL LETTER ETH (U+00D0)

  0x2015,	# HORIZONTAL BAR
    "\x81\x5C", # EM DASH (U+2014)

  0x2211,	# N-ARY SUMMATION (Windows CP932 0x8794)
    "\x83\xB0",	# GREEK CAPITAL LETTER SIGMA (U+03A3)

  0x2985,	# LEFT WHITE PARENTHESIS
    "\x81\xD4", # FULLWIDTH LEFT WHITE PARENTHESIS (U+FF5F)

  0x2986,	# RIGHT WHITE PARENTHESIS
    "\x81\xD5",	# FULLWIDTH RIGHT WHITE PARENTHESIS (U+FF60)

  0x3099,	# COMBINING KATAKANA-HIRAGANA VOICED SOUND MARK
    "\xDE",	# HALFWIDTH KATAKANA VOICED SOUND MARK (U+FF9E) // NFKC

  0x309A,	# COMBINING KATAKANA-HIRAGANA SEMI-VOICED SOUND MARK
    "\xDF",	# HALFWIDTH KATAKANA SEMI-VOICED SOUND MARK (U+FF9F) // NFKC

  0x4E44,	# CJK UNIFIED IDEOGRAPH-4E44 (JIS X 0212 16-17)
    "\x81\x59",	# IDEOGRAPHIC CLOSING MARK (U+3006)

  0x9B1C,	# CJK UNIFIED IDEOGRAPH-9B1C (for backward compatibility)
    "\xFC\x5A", # CJK UNIFIED IDEOGRAPH-9B1C

  0x9B1D,	# CJK UNIFIED IDEOGRAPH-9B1D (JIS X 0213:2000)
    "\xFC\x5A", # CJK UNIFIED IDEOGRAPH-9B1C

  0xFF3C,	# FULLWIDTH REVERSE SOLIDUS
    "\x81\x5F", # REVERSE SOLIDUS (U+005C)

  0xFF5E,	# FULLWIDTH TILDE
    "\x81\xB0",	# TILDE (U+007E)

  0xFFE0,	# FULLWIDTH CENT SIGN
    "\x81\x91",	# CENT SIGN (U+00A2)

  0xFFE1,	# FULLWIDTH POUND SIGN
    "\x81\x92",	# POUND SIGN (U+00A3)

  0xFFE2,	# FULLWIDTH NOT SIGN
    "\x81\xCA",	# NOT SIGN (U+00AC)

  0xFFE4,	# FULLWIDTH BROKEN BAR
    "\x85\x44",	# BROKEN BAR (U+00A6)
);

sub to_sjis0213_supplements { defined $_[0] && $Supplements{$_[0]} || '' }
sub to_sjis2004_supplements { defined $_[0] && $Supplements{$_[0]} || '' }

1;
__END__

=head1 NAME

ShiftJIS::X0213::MapUTF::Supplements - Supplemental Mapping from Unicode to Shift_JISX0213

=head1 SYNOPSIS

  use ShiftJIS::X0213::MapUTF;
  use ShiftJIS::X0213::MapUTF::Supplements;

  $sjis_str = utf8_to_sjis2004   (\&to_sjis2004_supplements, $utf8_str);
  $sjis_str = utf16_to_sjis2004  (\&to_sjis2004_supplements, $utf16_str);
  $sjis_str = utf16le_to_sjis2004(\&to_sjis2004_supplements, $utf16le_str);
  $sjis_str = utf16be_to_sjis2004(\&to_sjis2004_supplements, $utf16be_str);
  $sjis_str = utf32_to_sjis2004  (\&to_sjis2004_supplements, $utf32_str);
  $sjis_str = utf32le_to_sjis2004(\&to_sjis2004_supplements, $utf32le_str);
  $sjis_str = utf32be_to_sjis2004(\&to_sjis2004_supplements, $utf32be_str);
  $sjis_str = unicode_to_sjis2004(\&to_sjis2004_supplements, $unicode_str);

  $sjis_str = utf8_to_sjis0213   (\&to_sjis0213_supplements, $utf8_str);
  $sjis_str = utf16_to_sjis0213  (\&to_sjis0213_supplements, $utf16_str);
  $sjis_str = utf16le_to_sjis0213(\&to_sjis0213_supplements, $utf16le_str);
  $sjis_str = utf16be_to_sjis0213(\&to_sjis0213_supplements, $utf16be_str);
  $sjis_str = utf32_to_sjis0213  (\&to_sjis0213_supplements, $utf32_str);
  $sjis_str = utf32le_to_sjis0213(\&to_sjis0213_supplements, $utf32le_str);
  $sjis_str = utf32be_to_sjis0213(\&to_sjis0213_supplements, $utf32be_str);
  $sjis_str = unicode_to_sjis0213(\&to_sjis0213_supplements, $unicode_str);

=head1 DESCRIPTION

This module provides some supplemental mappings (fallbacks)
from Unicode to Shift_JISX0213, via a coderef.

=over 4

=item C<$sjis2004_char = to_sjis2004_supplements($unicode_codepoint)>

=item C<$sjis0213_char = to_sjis0213_supplements($unicode_codepoint)>

Returns a SJIS character (as a string) for some Unicode codepoints
unmapped to SJIS. Otherwise returns a null string.

B<NOTE:> C<to_sjis0213_supplements> is just an alias
for C<to_sjis2004_supplements>, then their mappings has no difference.

E.g. C<to_sjis2004_supplements(0x9B1D)> returns C<"\xFC\x5A">;
     C<to_sjis2004_supplements(0x00B5)> returns C<"\x83\xCA">;
     C<to_sjis0213_supplements(0x9B1C)> returns C<"\xFC\x5A">;
     C<to_sjis0213_supplements(0x00B5)> returns C<"\x83\xCA">.


=back

=head1 DISCLAIMER

This module is an B<experimental> release.
Propriety of mapping is not guaranteed.
Any of these supplemental mappings
may be added, modified, or removed in future.

=head1 AUTHOR

SADAHIRO Tomoyuki <SADAHIRO@cpan.org>

  Copyright(C) 2002-2007, SADAHIRO Tomoyuki. Japan. All rights reserved.

This module is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

=head1 SEE ALSO

=over 4

=item L<ShiftJIS::X0213::MapUTF>

=back

=cut


MapUTF.pm
package ShiftJIS::X0213::MapUTF;

require 5.006;

use strict;
use vars qw($VERSION $PACKAGE @ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);

require Exporter;
require DynaLoader;

@ISA = qw(Exporter DynaLoader);

$VERSION = '0.31';
$PACKAGE = 'ShiftJIS::X0213::MapUTF'; # __PACKAGE__

@EXPORT = qw(
    sjis0213_to_unicode  unicode_to_sjis0213
    sjis0213_to_utf8     utf8_to_sjis0213
    sjis0213_to_utf16le  utf16le_to_sjis0213
    sjis0213_to_utf16be  utf16be_to_sjis0213

    sjis2004_to_unicode  unicode_to_sjis2004
    sjis2004_to_utf8     utf8_to_sjis2004
    sjis2004_to_utf16le  utf16le_to_sjis2004
    sjis2004_to_utf16be  utf16be_to_sjis2004
);

%EXPORT_TAGS = (
    'unicode'  => [
	'sjis2004_to_unicode', 'unicode_to_sjis2004',
	'sjis0213_to_unicode', 'unicode_to_sjis0213',
    ],
    'utf8'     => [
	'sjis2004_to_utf8',    'utf8_to_sjis2004',
	'sjis0213_to_utf8',    'utf8_to_sjis0213',
    ],
    'utf16'    => [
	                       'utf16_to_sjis2004',
	                       'utf16_to_sjis0213',
    ],
    'utf16le'  => [
	'sjis2004_to_utf16le', 'utf16le_to_sjis2004',
	'sjis0213_to_utf16le', 'utf16le_to_sjis0213',
    ],
    'utf16be'  => [
	'sjis2004_to_utf16be', 'utf16be_to_sjis2004',
	'sjis0213_to_utf16be', 'utf16be_to_sjis0213',
    ],
    'utf32'    => [
	                       'utf32_to_sjis2004',
	                       'utf32_to_sjis0213',
    ],
    'utf32le'  => [
	'sjis2004_to_utf32le', 'utf32le_to_sjis2004',
	'sjis0213_to_utf32le', 'utf32le_to_sjis0213',
    ],
    'utf32be'  => [
	'sjis2004_to_utf32be', 'utf32be_to_sjis2004',
	'sjis0213_to_utf32be', 'utf32be_to_sjis0213',
    ],
);

@EXPORT_OK = map @$_, values %EXPORT_TAGS;
$EXPORT_TAGS{all}  = [ @EXPORT_OK ];

bootstrap ShiftJIS::X0213::MapUTF $VERSION;

1;
__END__

=head1 NAME

ShiftJIS::X0213::MapUTF - conversion between Shift_JIS-2004/Shift_JISX0213 and Unicode

=head1 SYNOPSIS

    use ShiftJIS::X0213::MapUTF;

    # for Shift_JIS-2004
    $utf16be_string  = sjis2004_to_utf16be($sjis2004_string);
    $sjis2004_string = utf16be_to_sjis2004($utf16be_string);

    # for Shift_JISX0213
    $utf16be_string  = sjis0213_to_utf16be($sjis0213_string);
    $sjis0213_string = utf16be_to_sjis0213($utf16be_string);

=head1 DESCRIPTION

This module provides functions to convert
from Shift_JIS-2004 (specified by JIS X 0213:2004) to Unicode, and vice versa.

For backward compatibility, this module also provides functions to convert
from Shift_JISX0213 (specified by JIS X 0213:2000) to Unicode, and vice versa.

For convenience, "SJIS-X" is used to refer to
both Shift_JIS-2004 and Shift_JISX0213 hereafter.

The following 10 JIS Kanji characters are added in JIS X 0213:2004.
These mappings are used only for Shift_JIS-2004,
and not for Shift_JISX0213.

   sjis2004     unicode 3.2.0

    0x879F        U+4FF1
    0x889E        U+525D
    0x9873        U+20B9F
    0x989E        U+541E
    0xEAA5        U+5653
    0xEFF8        U+59F8
    0xEFF9        U+5C5B
    0xEFFA        U+5E77
    0xEFFB        U+7626
    0xEFFC        U+7E6B

=head2 Conversion from SJIS-X to Unicode

If the first parameter is a reference,
that is used for coping with SJIS-X characters
unmapped to Unicode, C<SJIS_CALLBACK>.
(any reference will not allowed as C<STRING>.)

If C<SJIS_CALLBACK> is given, C<STRING> is
the second parameter; otherwise the first.

If C<SJIS_CALLBACK> is not specified,
SJIS-X characters unmapped to Unicode are silently deleted
and illegal bytes are skipped by one byte.
(as if a coderef constantly returning null string, C<sub {''}>,
is passed as C<SJIS_CALLBACK>.)

Currently, only coderefs are allowed as C<SJIS_CALLBACK>.
A string returned from C<SJIS_CALLBACK> is inserted
in place of the unmapped character or the illegal byte.

A coderef as C<SJIS_CALLBACK> is called with one or more arguments.

If illegal byte appears (i.e. a leading byte C<[0x81..0x9F, 0xE0..0xFC]>
without trailing byte (C<[0x40..0x7E, 0x80..0xFC]>), or a reserved byte
(C<[0x80, 0xA0, 0xF0..0xFF]>), the first argument is C<undef>
and the second argument is an unsigned integer representing the byte.

If an unmapped character appears, the first argument is
a defined string representing a character.

Example

    my $sjis_callback = sub {
        my ($char, $byte) = @_;
        return function($char) if defined $char;
        die sprintf "illegal byte 0x%02x", $byte;
    };

In the example above, C<$char> may be C<"\xfc\xfc">, etc.

The return value of C<SJIS_CALLBACK> must be legal in the target format.
E.g. never use with C<sjis2004_to_utf16be()> a callback that returns UTF-8.
I.e. you should prepare C<SJIS_CALLBACK> for each UTF.

=over 4

=item C<sjis2004_to_utf8([SJIS_CALLBACK,] STRING)>

Converts Shift_JIS-2004 to UTF-8

=item C<sjis2004_to_unicode([SJIS_CALLBACK,] STRING)>

Converts Shift_JIS-2004 to Unicode
(Perl's internal format, flagged with C<SVf_UTF8>,
see F<perlunicode>)

=item C<sjis2004_to_utf16le([SJIS_CALLBACK,] STRING)>

Converts Shift_JIS-2004 to UTF-16LE.

=item C<sjis2004_to_utf16be([SJIS_CALLBACK,] STRING)>

Converts Shift_JIS-2004 to UTF-16BE.

=item C<sjis2004_to_utf32le([SJIS_CALLBACK,] STRING)>

Converts Shift_JIS-2004 to UTF-32LE.

=item C<sjis2004_to_utf32be([SJIS_CALLBACK,] STRING)>

Converts Shift_JIS-2004 to UTF-32BE.

=item C<sjis0213_to_utf8([SJIS_CALLBACK,] STRING)>

Converts Shift_JISX0213 to UTF-8

=item C<sjis0213_to_unicode([SJIS_CALLBACK,] STRING)>

Converts Shift_JISX0213 to Unicode
(Perl's internal format, flagged with C<SVf_UTF8>,
see F<perlunicode>)

=item C<sjis0213_to_utf16le([SJIS_CALLBACK,] STRING)>

Converts Shift_JISX0213 to UTF-16LE.

=item C<sjis0213_to_utf16be([SJIS_CALLBACK,] STRING)>

Converts Shift_JISX0213 to UTF-16BE.

=item C<sjis0213_to_utf32le([SJIS_CALLBACK,] STRING)>

Converts Shift_JISX0213 to UTF-32LE.

=item C<sjis0213_to_utf32be([SJIS_CALLBACK,] STRING)>

Converts Shift_JISX0213 to UTF-32BE.

=back

=head2 Conversion from Unicode to SJIS-X

If the first parameter is a reference,
that is used for coping with Unicode characters
unmapped to SJIS-X, C<UNICODE_CALLBACK>.
(any reference will not allowed as C<STRING>.)

If C<UNICODE_CALLBACK> is given, C<STRING> is
the second parameter; otherwise the first.

If C<UNICODE_CALLBACK> is not specified, SJIS-X characters
unmapped to Unicode are silently deleted
and partial bytes are skipped by one byte.
(as if a coderef constantly returning null string, C<sub {''}>
is passed as C<UNICODE_CALLBACK>.)

Currently, only coderefs are allowed as C<UNICODE_CALLBACK>.
A string returned from the coderef is inserted
in place of the unmapped character.

A coderef as C<UNICODE_CALLBACK> is called with one or more arguments.
If the unmapped character is a partial character (an illegal byte),
the first argument is C<undef>
and the second argument is an unsigned integer representing the byte.
If not partial, the first argument is an unsigned interger
representing a Unicode code point.

For example, characters unmapped to SJIS-X are
converted to numerical character references for HTML 4.01.

    sub toHexNCR {
        my ($char, $byte) = @_;
        return sprintf("&#x%x;", $char) if defined $char;
        die sprintf "illegal byte 0x%02x", $byte;
    }

    $sjis2004 = utf8_to_sjis2004   (\&toHexNCR, $utf8_string);
    $sjis2004 = unicode_to_sjis2004(\&toHexNCR, $unicode_string);
    $sjis2004 = utf16le_to_sjis2004(\&toHexNCR, $utf16le_string);

The return value of C<UNICODE_CALLBACK> must be legal in Shift_JIS-2004.

=over 4

=item C<utf8_to_sjis2004([UNICODE_CALLBACK,] STRING)>

Converts UTF-8 to Shift_JIS-2004.

=item C<unicode_to_sjis2004([UNICODE_CALLBACK,] STRING)>

Converts Unicode to Shift_JIS-2004.

This B<Unicode> is in the Perl's internal format (see F<perlunicode>).
If C<SVf_UTF8> is not turned on,
C<STRING> is upgraded as an ISO 8859-1 (latin1) string.

=item C<utf16_to_sjis2004([UNICODE_CALLBACK,] STRING)>

Converts UTF-16 (with or w/o C<BOM>) to Shift_JIS-2004.

=item C<utf16le_to_sjis2004([UNICODE_CALLBACK,] STRING)>

Converts UTF-16LE to Shift_JIS-2004.

=item C<utf16be_to_sjis2004([UNICODE_CALLBACK,] STRING)>

Converts UTF-16BE to Shift_JIS-2004.

=item C<utf32_to_sjis2004([UNICODE_CALLBACK,] STRING)>

Converts UTF-32 (with or w/o C<BOM>) to Shift_JIS-2004.

=item C<utf32le_to_sjis2004([UNICODE_CALLBACK,] STRING)>

Converts UTF-32LE to Shift_JIS-2004.

=item C<utf32be_to_sjis2004([UNICODE_CALLBACK,] STRING)>

Converts UTF-32BE to Shift_JIS-2004.

=item C<utf8_to_sjis0213([UNICODE_CALLBACK,] STRING)>

Converts UTF-8 to Shift_JISX0213.

=item C<unicode_to_sjis0213([UNICODE_CALLBACK,] STRING)>

Converts Unicode to Shift_JISX0213.

This B<Unicode> is in the Perl's internal format (see F<perlunicode>).
If C<SVf_UTF8> is not turned on,
C<STRING> is upgraded as an ISO 8859-1 (latin1) string.

=item C<utf16_to_sjis0213([UNICODE_CALLBACK,] STRING)>

Converts UTF-16 (with or w/o C<BOM>) to Shift_JISX0213.

=item C<utf16le_to_sjis0213([UNICODE_CALLBACK,] STRING)>

Converts UTF-16LE to Shift_JISX0213.

=item C<utf16be_to_sjis0213([UNICODE_CALLBACK,] STRING)>

Converts UTF-16BE to Shift_JISX0213.

=item C<utf32_to_sjis0213([UNICODE_CALLBACK,] STRING)>

Converts UTF-32 (with or w/o C<BOM>) to Shift_JISX0213.

=item C<utf32le_to_sjis0213([UNICODE_CALLBACK,] STRING)>

Converts UTF-32LE to Shift_JISX0213.

=item C<utf32be_to_sjis0213([UNICODE_CALLBACK,] STRING)>

Converts UTF-32BE to Shift_JISX0213.

=back

=head2 Export

B<By default:>

    sjis2004_to_utf8     utf8_to_sjis2004
    sjis2004_to_utf16le  utf16le_to_sjis2004
    sjis2004_to_utf16be  utf16be_to_sjis2004
    sjis2004_to_unicode  unicode_to_sjis2004

    sjis0213_to_utf8     utf8_to_sjis0213
    sjis0213_to_utf16le  utf16le_to_sjis0213
    sjis0213_to_utf16be  utf16be_to_sjis0213
    sjis0213_to_unicode  unicode_to_sjis0213

B<On request:>

    sjis2004_to_utf32le  utf32le_to_sjis2004
    sjis2004_to_utf32be  utf32be_to_sjis2004
                         utf16_to_sjis2004 [*]
                         utf32_to_sjis2004 [*]

    sjis0213_to_utf32le  utf32le_to_sjis0213
    sjis0213_to_utf32be  utf32be_to_sjis0213
                         utf16_to_sjis0213 [*]
                         utf32_to_sjis0213 [*]

[*] Their counterparts C<sjis2004_to_utf16()>, C<sjis2004_to_utf32()>,
C<sjis0213_to_utf16()> and C<sjis0213_to_utf32()>
are not implemented yet. They need more investigation
on return values from C<SJIS_CALLBACK>...
(concatenation needs recognition of and coping with C<BOM>)

=head1 BUGS

On mapping between SJIS-X and Unicode used in this module,
notice that:

=over 4

=item *

0xFC5A in both Shift_JIS-2004 and Shift_JISX0213
is mapped to U+9B1C according to JIS X 0213:2004,
although JIS X 0213:2000 mapped it to U+9B1D.

=item *

The following 25 JIS Non-Kanji characters are not included in Unicode 3.2.0.
So they are mapped to each 2 characters in Unicode.
These mappings are done round-trippedly for *one SJIS-X character*.
Then round-trippedness for a SJIS-X *string* is broken.
(E.g. SJIS-X <0x8663> and <0x857B, 0x867B> both are mapped
to <U+00E6, U+0300>; but <U+00E6, U+0300> is mapped only to SJIS-X <0x8663>.)

    SJIS-X     Unicode 3.2.0    # Name by JIS X 0213:2004

    0x82F5    <U+304B, U+309A> # [HIRAGANA LETTER BIDAKUON NGA]
    0x82F6    <U+304D, U+309A> # [HIRAGANA LETTER BIDAKUON NGI]
    0x82F7    <U+304F, U+309A> # [HIRAGANA LETTER BIDAKUON NGU]
    0x82F8    <U+3051, U+309A> # [HIRAGANA LETTER BIDAKUON NGE]
    0x82F9    <U+3053, U+309A> # [HIRAGANA LETTER BIDAKUON NGO]
    0x8397    <U+30AB, U+309A> # [KATAKANA LETTER BIDAKUON NGA]
    0x8398    <U+30AD, U+309A> # [KATAKANA LETTER BIDAKUON NGI]
    0x8399    <U+30AF, U+309A> # [KATAKANA LETTER BIDAKUON NGU]
    0x839A    <U+30B1, U+309A> # [KATAKANA LETTER BIDAKUON NGE]
    0x839B    <U+30B3, U+309A> # [KATAKANA LETTER BIDAKUON NGO]
    0x839C    <U+30BB, U+309A> # [KATAKANA LETTER AINU CE]
    0x839D    <U+30C4, U+309A> # [KATAKANA LETTER AINU TU]
    0x839E    <U+30C8, U+309A> # [KATAKANA LETTER AINU TO]
    0x83F6    <U+31F7, U+309A> # [KATAKANA LETTER AINU P]
    0x8663    <U+00E6, U+0300> # [LATIN SMALL LETTER AE WITH GRAVE]
    0x8667    <U+0254, U+0300> # [LATIN SMALL LETTER OPEN O WITH GRAVE]
    0x8668    <U+0254, U+0301> # [LATIN SMALL LETTER OPEN O WITH ACUTE]
    0x8669    <U+028C, U+0300> # [LATIN SMALL LETTER TURNED V WITH GRAVE]
    0x866A    <U+028C, U+0301> # [LATIN SMALL LETTER TURNED V WITH ACUTE]
    0x866B    <U+0259, U+0300> # [LATIN SMALL LETTER SCHWA WITH GRAVE]
    0x866C    <U+0259, U+0301> # [LATIN SMALL LETTER SCHWA WITH ACUTE]
    0x866D    <U+025A, U+0300> # [LATIN SMALL LETTER HOOKED SCHWA WITH GRAVE]
    0x866E    <U+025A, U+0301> # [LATIN SMALL LETTER HOOKED SCHWA WITH ACUTE]
    0x8685    <U+02E9, U+02E5> # [RISING SYMBOL]
    0x8686    <U+02E5, U+02E9> # [FALLING SYMBOL]

=back

=head1 AUTHOR

SADAHIRO Tomoyuki <SADAHIRO@cpan.org>

  Copyright(C) 2002-2007, SADAHIRO Tomoyuki. Japan. All rights reserved.

This module is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

=head1 SEE ALSO

=over 4

=item JIS X 0213:2000/Amd1:2004

7-bit and 8-bit double byte coded extended KANJI sets
for information interchange

=item Japanese Industrial Standards Committee (JISC)

L<http://www.jisc.go.jp/>

=item Japanese Standards Association (JSA)

L<http://www.jsa.or.jp/>

=item Unihan database (Unicode version: 3.2.0) by Unicode (c)

L<http://www.unicode.org/Public/UNIDATA/Unihan.txt>

=item JIS KANJI JITEN, the revised edition

edited by Shibano, published by Japanese Standards Association,
2002, Tokyo [ISBN4-542-20129-5]

=item L<ShiftJIS::CP932::MapUTF>

conversion between Microsoft Windows CP-932 and Unicode

(CP932-Unicode mapping is different with Shift_JIS-2004-Unicode mapping,
but what you desire may be the former.)

=back

=cut
